Enabled the option to init graphics outside the constructor.

dtorban opened the pull request:
The current implementation of MinGfx doesn't allow for offline testing because if we inherit from GraphicApp it will try to use a graphics context. I moved this code into an initWindow() method that is called when the Run() method is called. This way it will only try to draw graphics when the graphics loop is started. I also added a virtual method "void InitGraphics()" for initializing graphics code (this is for graphics code like buttons, sliders, nanogui stuff, etc...)

In order to maintain backwards compatibility, everything will work like it used to, however, when we are ready to force the use of InitGraphics(), I added a CMake option FORCE_INIT_GFX_NO_CONSTRUCTOR.  By default, this is set to OFF.  For the automated command line test scripts, however, it will be set to ON.  Someday, perhaps we can remove the work around completely.

dankeefe commented:
Thanks Dan,
This looks good.  Can you make 1 change?  I'm anticipating there might be some confusion about what to use InitGraphics() for and what to use InitOpenGL() for.  Can you rename InitGraphics() to InitNanoGUI()? Then, we will just make it clear that that is where to do your NanoGUI stuff and that this is distinct from the opengl initialization.  Maybe after the class is over, we'll collapse both functions into a single InitGraphics().

dtorban commented:
Dan, I renamed InitGraphics to InitNanoGUI and also changed FORCE_INIT_GFX_NO_CONSTRUCTOR int ALLOW_INIT_GFX_CTX_IN_CONSTRUCTOR.  Someday we can just turn this off.

I also made all the glfw specific functions like:
- virtual bool IsKeyDown(int key);
- ...
- virtual int window_width();

This way we can override them for testing.
